lib/filesystem.nix

```diff
@@ -154,4 +166,162 @@ in
+
+      packagesFromSubdirectory = subdirectory:
+        let
+          supportedDirectoryEntries =
+            filterAttrs supported (readDir subdirectory);
+
+          directoryEntryToPackageEntries = basename: type:
+            let
+              path = subdirectory + "/${basename}";
+            in
+            {
+              name = replaceStrings [ ".nix" ] [ "" ] basename;
+
+              value =
+                if type == "regular"
+                then path
+                else if type == "directory"
+                then
+                  let
+                    defaultPackagePath = path + "/package.nix";
+                  in
+                  if pathExists defaultPackagePath
+                  then defaultPackagePath
+                  else packagesFromSubdirectory path
+                else
+                  abort
+                    ''
+                      packagesFromDirectory: Unsupported file type
+
+                      File type: ${type}
+
+                      Path: ${path}
+                    '';
+            };
+        in
+        mapAttrs' directoryEntryToPackageEntries supportedDirectoryEntries;
+    in
+    packagesFromSubdirectory directory;
+
+  /*
+    Transform a directory tree containing package files suitable for
+    `callPackage` into a matching nested attribute set of derivations.
+
+    For a directory tree like this:
+
+    ```
+    my-packages
+    ├── a.nix
+    ├── b.nix
+    ├── c
+    │  ├── my-extra-feature.patch
+    │  ├── package.nix
+    │  └── support-definitions.nix
+    └── my-namespace
+       ├── d.nix
+       ├── e.nix
+       └── f
+          └── package.nix
+    ```
+
+    `packagesFromDirectory pkgs.callPackage ./my-packages` will return an attribute set like this:
+
+    ```nix
+    {
+      a = pkgs.callPackage ./my-packages/a.nix { };
+      b = pkgs.callPackage ./my-packages/b.nix { };
+      c = pkgs.callPackage ./my-packages/c/package.nix { };
+      my-namespace = {
+        d = pkgs.callPackage ./my-packages/my-namespace/d.nix { };
+        e = pkgs.callPackage ./my-packages/my-namespace/e.nix { };
+        f = pkgs.callPackage ./my-packages/my-namespace/f/package.nix { };
+      };
+    }
+    ```
+
+    See the [`packagesFromDirectory`](#sec-packagesfromdirectory) section in
+    the manual for a guide on using `packagesFromDirectory`, and see the
+    [`packageFilesFromDirectory`](#function-library-lib.filesystem.packageFilesFromDirectory)
+    documentation for details on how the directory tree is traversed.
+
+    Example:
+      packagesFromDirectory pkgs.callPackage ./my-packages
+      => { ... }
+
+      lib.makeScope pkgs.newScope (
+        self: packagesFromDirectory self.callPackage ./my-packages
+      )
+      => { ... }
+
+    Type:
+      packagesFromDirectory :: (Path -> AttrSet -> Package) -> Path -> AttrSet
+  */
+  packagesFromDirectory =
+    # `pkgs.callPackage`
+    callPackage:
+    # The directory to read package files from
+    directory:
```

Comment:
The main thing I'll suggest is to make this take an attribute set instead of position arguments:

```lib
{ callPackage, directory }:
```

… that way if we need to add new optional arguments (with sensible defaults) we don't have to break backwards-compatibility.  This would also be consistent with how `pkgs.haskell.lib.packagesFromDirectory` works (it also takes an attribute set)

Reply:
Oh, and also, have the attribute set ignore extra arguments (for forward compatibility), also the same as `pkgs.haskell.lib.packagesFromDirectory`:

```nix
{ callPackage, directory, ... }:
```

Reply:
Done.

Reply:
I'd be fine removing the `...`.  I'm not *too* attached to it.

Reply:
+1 for dropping `...`!